### ISSUE TITLE
Add arbitrary fieldname in set_float_info

### ISSUE THREAD

**Issue by guyko81** [dmlc/xgboost#5990]
In #2255 the same question was asked but the issue was closed with no resolution. 
Since then at least 2 new metainfo was added to the package: labels_lower_bound_ and labels_upper_bound_. 
[https://xgboost.readthedocs.io/en/stable/dev/classxgboost_1_1MetaInfo.html](https://xgboost.readthedocs.io/en/stable/dev/classxgboost_1_1MetaInfo.html)
These new metainfo fields let xgboost to calculate survival models.

In my case I need to add additional logic to the calculation in my custom loss. I can trick the system by passing my data through the labels_lower_bound and labels_upper_bound, but it's not nice at all and I was wondering if I needed a third parameter for my calculation. 

Is it possible to add any field name and reference them in the custom loss with their name?

The documentation doesn't mention any restriction on field names.
https://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.DMatrix.set_float_info

**Comment by trivialfis** [comment]
Could you please provide an example on what you are trying to do?  I'm confused why DMatrix has to be used as a container for your other information.

**Comment by guyko81** [comment]
I'm using a row level information like the one in AFT survival model
[https://xgboost.readthedocs.io/en/latest/tutorials/aft_survival_analysis.html](https://xgboost.readthedocs.io/en/latest/tutorials/aft_survival_analysis.html)

The reason I'm doing this is that I'm making a custom loss function that is similar to categorical crossentropy but sometimes I need to fill more than 1 column after the OneHotEncoding to properly optimise my loss. The target can be in many columns (categories) with different probabilities, and I'm aware of it in the training data.

Unfortunately I'm not able to add multiple target columns neither. That way I could pass any arbitrary info (e.g. the whole probability distribution over the categories).

**Comment by trivialfis** [comment]
Can you do it with closure?

``` python

discrete_values = np.random.randn(rows, cols)   # <-- The extra info

def my_obj(predt, dtrain):
  y = dtrain.get_label()
  new_predt = do_something(predt, discrete_values)
  grad = my_grad_func(y, new_predt)
  hess = my_hess_func(y, new_predt)
  return grad, hess
```

**Comment by guyko81** [comment]
I think this solution wouldn't work in case of subsample<1 due to random behaviour of row selection.

**Comment by trivialfis** [comment]
subsample is used for constructing tree, not for objective.  Objective always computes gradient for all the samples, the tree updater filter out those not being sampled.

**Comment by guyko81** [comment]
Interesting, didn't know. Thanks for letting me know.
In that case I don't see any reason not to use your proposal! Thanks 👍 

Just can you please ensure me that the order is kept as well? Not like I'm expecting other, just want to be sure.

**Comment by trivialfis** [comment]
Glad that I can help.  The order is preserved.
